content/2018-10-31-xsimd.rst

```diff
@@ -0,0 +1,271 @@
+
+.. code-block:: sh
+
+    CC=gcc CXX=g++ pythran -march=native -O3 normalize_complex_arr.py
+    python -m perf timeit -s 'import numpy as np; np.random.seed(0); N = 100000; x = np.random.random(N) + 1j *  np.random.random(N); from normalize_complex_arr import normalize_complex_arr' 'normalize_complex_arr(x)'
+    .....................
+    Mean +- std dev: 2.84 ms +- 0.01 ms
+
+
+And with vectorization on (note that the calls to ``np.min()`` do not involve complex numbers though.
+
+.. code-block:: sh
+
+    CC=gcc CXX=g++ pythran -march=native -O3 make_decision.py -DUSE_XSIMD
+    python -m perf timeit -s 'import numpy as np; np.random.seed(0); N = 100000; x = np.random.random(N) + 1j *  np.random.random(N); from normalize_complex_arr import normalize_complex_arr' 'normalize_complex_arr(x)'
+    .....................
+    Mean +- std dev: 723 us +- 14 us
+
+Cool! Speedup for complex! For the record, the numpy version already ran at roughly ``3.19 ms +- 0.02 ms``.
+
+Scalar Version
+==============
+
+That's probably a detail for many xsimd users, but thanks to this cooperation,
+xsimd now exposes a scalar version of all its mathematical function inside the
+``xsimd::`` namespace. That way one can write higher level functions based on
+xsimd, and they would work for both scalar and vector version:
+
+.. code-block:: c++
+
+    template<class T>
+    T euclidian_distance_squared(T x, T y)
+    {
+        auto tmp = xsimd::hypot(x, y);
+        return tmp * tmp;
+    }
+
+In the context of Pythran, this makes the expression template engine easier to
+write. Good point.
+
+
+Compilation Time
+================
+
+Pythran is an *Ahead of Time* compiler, so compilation time is generally not a
+good metric. But there's one situation where it matters to me: Continuous
+Integration. Because Travis has time limits, the faster we compile, the more
+test we can pass! As Pythran validates for Python2 and Python3, for Clang and
+GCC, with and without SIMD, with and without OpenMP, that's a lot of
+configuration to test. Roughly... 20hours of cumulated tests actually, see
+`this recent build
+<https://travis-ci.com/serge-sans-paille/pythran/builds/89663340>`_ for
+instance.
+
+In pre-xsimd setting, compiling the above  ``arc_distance.py`` file in simd mode is relatively slow:
+
+.. code-block:: sh
+
+    time pythran -O3 -march=native normalize_complex_arr.py -E # generate the .cpp
+    0.91s user 0.28s system 130% cpu 0.908 total
+    time pythran -O3 -march=native normalize_complex_arr.cpp -DUSE_BOOST_SIMD
+    12.10s user 0.79s system 102% cpu 12.616 total
+
+
+With xsimd, it's slightly faster (no boost dependencies, and less C++ magic).
+
+.. code-block:: sh
+
+    time pythran -O3 -march=native arc_distance.cpp -DUSE_XSIMD
+    10.32s user 0.65s system 102% cpu 10.688 total
+
+Arguably, it's still slow. But a bit less then what it uses to be, and the reason also lies in the generated c++, not only in the vectorization engine:
```

Comment:
This explanation is a bit hard to follow. I think grouping the timing of the initial cpp file generation and boost.simd parts at the beginning does more harm than good on the clarity front. I think it would be easier to understand in this order/grouping:

- Generation of cpp and raw pythran compile time ("reference")
- Boost.SIMD version ("before")
- xsimd version ("after")